pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -59,73 +85,343 @@ contract LiquidityBootstrappingPool is BaseWeightedPool, ReentrancyGuard {
             owner
         )
     {
-        uint256 numTokens = tokens.length;
-        _require(numTokens <= _MAX_LBP_TOKENS, Errors.MAX_TOKENS);
+        uint256 totalTokens = tokens.length;
+        _require(totalTokens <= _MAX_LBP_TOKENS, Errors.MAX_TOKENS);
+
+        InputHelpers.ensureInputLengthMatch(totalTokens, normalizedWeights.length);
 
-        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length);
+        bytes32 poolState;
 
-        // Ensure each normalized weight is above the minimum and find the token index of the maximum weight
+        // Ensure valid weights
         uint256 normalizedSum = 0;
-        uint256 maxWeightTokenIndex = 0;
-        uint256 maxNormalizedWeight = 0;
-        for (uint8 i = 0; i < numTokens; i++) {
+        for (uint8 i = 0; i < totalTokens; i++) {
             uint256 normalizedWeight = normalizedWeights[i];
             _require(normalizedWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
 
-            _setNormalizedWeight(normalizedWeight, i);
+            // Insert "start weights" and "end weights" into poolState
+            // Before the first update, start=end, which avoids initialization edge cases
```

Comment:
Well, there is no weight update until they call updateWeightsGradually (which they theoretically might never do). I also found I didn't really need to initialize the times (it just checks if they're equal, so 0 works as well as any other value, and saves the gas/bytecode of writing values that are never used).

Reply:
Sure, but imagine you develop an appliation using events, or e.g. do a subgraph. You wouldn't be able to compute the weights from events alone because there is none for the initial weights: you'd require some initial state. 

This is why we emit events in the constructor for all of these things, it makes those applications much easier to develop, and the cost of emitting the events/writing to storage are negligible compared to deploying the Pool. Same with setting the start and end times: construction is equivalent to a zero-duration weight change, and making it be that way means there is no edge-case of 'zero start time' that needs to be handled.

Reply:
Yes - this plus refactoring the initialize weight change thing makes it very clean!